[#41939] Email delivery not working on qa-edge

cbliard opened the pull request:
https://community.openproject.org/work_packages/41939

because of env vars without `OPENPROJECT_` prefix being ignored since the merge of #10296 "harmonize settings & configuration"
As SMTP settings were defined with `SMTP_xxx` env vars, they got ignored.

Settings should work with all those variations:
- `OPENPROJECT_SMTP__ENABLE__STARTTLS__AUTO=true` (currently the only one supported)
- `OPENPROJECT_SMTP_ENABLE_STARTTLS_AUTO=true`
- `SMTP_ENABLE_STARTTLS_AUTO=true` (with a deprecation warning)

cbliard committed:
load settings from variations of env variables

could load value from:
- OPENPROJECT_SOME__KEY__NAME=value
can now also load value from:
- OPENPROJECT_SOME_KEY_NAME=value
- SOME_KEY_NAME=value

A deprecation message is logged if used without the OPENPROJECT_ prefix.

ulferts reviewed (changes requested):
Works to handle all of the ways defined in this PR correctly. There are two cases I think are not ideal:

1) It is possible to define a setting multiple times now e.g. 

```
OPENPROJECT_ACTIVITY__DAYS__DEFAULT=20 OPENPROJECT_ACTIVITY_DAYS_DEFAULT=10 ACTIVITY_DAYS_DEFAULT=15 rails s
```

and one has to read to code to know which one will win. 

2) With the introduction of the non duplicate underscore named variants defining an env var has become easier for most of the cases. But for hashed definitions, it has become more complicated. Because only there, an env variable still has to be defined with two consecutive underscores:

```
OPENPROJECT_REPOSITORY__CHECKOUT__DATA_GIT_ENABLED=1 rails s
```

will work while

```
OPENPROJECT_REPOSITORY_CHECKOUT_DATA_GIT_ENABLED=1 rails s
```

won't. That is because only the 'legacy' style of writing is supported.

Even if that is fixed, the inconsistency of having to have two consecutive underscores in case of a nested hash key will remain (e.g.: `OPENPROJECT_REPOSITORY__CHECKOUT__DATA_GIT__BLUBS_ENABLED`) but that is probably unavoidable unless a way is found so that the structure of a definition hash and by that the available keys can be defined. 

Of those two cases, only the second one really needs to be addressed before merging in my opinion to avoid confusion.

cbliard commented:
> Works to handle all of the ways defined in this PR correctly. There are two cases I think are not ideal:
> 
> 1. It is possible to define a setting multiple times now e.g.
> 
> ```
> OPENPROJECT_ACTIVITY__DAYS__DEFAULT=20 OPENPROJECT_ACTIVITY_DAYS_DEFAULT=10 ACTIVITY_DAYS_DEFAULT=15 rails s
> ```
> 
> and one has to read to code to know which one will win.

Yes. Should it be documented at the same location where it is explained that env vars win over config file? Or should we have a warning logged whenever a setting location is overriding another location?

Also I thought that keeping the information about where the setting comes from (`:db`, `:default`, `:env`, `:file`) in a `from` field of the `Definition` instance could be useful, so the information can be diagnosed, but thought it would be out of scope so I left the idea.

> 2. With the introduction of the non duplicate underscore named variants defining an env var has become easier for most of the cases. But for hashed definitions, it has become more complicated. Because only there, an env variable still has to be defined with two consecutive underscores:
> 
> ```
> OPENPROJECT_REPOSITORY__CHECKOUT__DATA_GIT_ENABLED=1 rails s
> ```
> 
> will work while
> 
> ```
> OPENPROJECT_REPOSITORY_CHECKOUT_DATA_GIT_ENABLED=1 rails s
> ```
> 
> won't. That is because only the 'legacy' style of writing is supported.
> 
> Even if that is fixed, the inconsistency of having to have two consecutive underscores in case of a nested hash key will remain (e.g.: `OPENPROJECT_REPOSITORY__CHECKOUT__DATA_GIT__BLUBS_ENABLED`) but that is probably unavoidable unless a way is found so that the structure of a definition hash and by that the available keys can be defined.
> 
> Of those two cases, only the second one really needs to be addressed before merging in my opinion to avoid confusion.

Same, I got a working implementation supporting both cases at some point with double underscores being the nesting separator. Here too I thought that it might be too complicated and not needed by anyone. Maybe supporting one level of nesting with single underscore syntax could be enough.

Thoughts @ulferts?

cbliard commented:
After discussing with Jens, `OPENPROJECT_REPOSITORY_CHECKOUT_DATA_GIT_ENABLED=1` and `OPENPROJECT_REPOSITORY__CHECKOUT__DATA_GIT_ENABLED=1` should behave identically.

I'll implement it.

cbliard committed:
handle env vars overriding hash setting with single underscore

setting OPENPROJECT_REPOSITORY_CHECKOUT_DATA_GIT_MINIMUM__VERSION=42 will do the same as OPENPROJECT_REPOSITORY__CHECKOUT__DATA_GIT_MINIMUM__VERSION=42: it will do Setting.repository_checkout_data['git']['minimum_data'] = 1

Please use OPENPROJECT_REPOSITORY_CHECKOUT_DATA='{git: {minimum_version: 42}}' instead, it is more readable.

cbliard commented:
After discussion with Oliver, we could do that:

- [x] Keep the current approach with nesting (_) and underscore (__) for backwards compatibility
- [x] Add support for values with single underscore syntax as Jens suggested, and keep the hash part as before
- [ ] Deprecate all ENVs that are passed in this way
  - [x] Use `Rails.logger.warn` for deprecation. Warning: some values may be secrets so the deprecation message must not output the value.
  - [ ] Update docs https://www.openproject.org/docs/installation-and-operations/configuration/ and also https://www.openproject.org/docs/installation-and-operations/configuration/environment/
- [ ] Promote the use of inline YML in docs and remove the nesting in 13.0
- optionally: Provide some way of converting a YML file to the respective ENVs
  - so users can use a configuration.yml to start with, and then convert it to ENVs to be used in a docker context

some examples:

Setting
```
OPENPROJECT_MULTITENANCY_ZACERO_ACCESS__TOKEN_EXPIRATION__TIME=3600
```
will result in
```
OpenProject::Configuration.multitenancy
# { zazero: { access_token: { expiration_time: 3600 } } }
```
And a pricing example
```
OPENPROJECT_SUBSCRIPTIONS_PLANS_60_PROFESSIONAL_US=enterprise-bim-cloud---professional---usd---5-year
```

cbliard committed:
log deprecation warning when using env var names to set nested hash setting

cbliard committed:
refactor: make it consistent with other WEB settings

cbliard committed:
stringify hash settings keys and support OPENPROJECT_2FA env var

updated code to look up the string key instead of the symbol key

introduce a 'with settings' rspec shared context

cbliard committed:
make nested hash env names work with env_alias for Setting

cbliard commented:
Seen with Oliver: we'll do the deprecation, doc updates and promotion of yaml/json in env vars in another PR.

cbliard committed:
remove deprecation warning when using nested hash in env var names

machisuji commented on config/constants/settings/definition.rb:
```diff
@@ -278,32 +288,80 @@ def unescape_underscores(path_segment)
+        ].compact
+      end
+
+      def env_name_legacy(definition)
```
Since this is not/no longer deprecated I'd prefer this to be named differently. How about `env_name_nested`?

cbliard replied:
good point.

I'm just seeing your comment, and Oliver already took care of it :)

machisuji reviewed (changes requested):
Thanks @cbliard, looking good. I have just one minor nit-pick.
